### PR TITLE
#23 working on lexer module

### LINKED ISSUE #23: Init lexer module

### PR DESCRIPTION
**Pull request by alex216**
fix #23 

- **Add realloc for lexer**
- **Add push_back_char for lexer**
- **Add store_to_buffer_statefully**
- **Add base lexer**

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive lexer component for improved input parsing and tokenization.
	- Added functionality for processing comments, operators, quotes, and blank spaces within the lexer.
	- Implemented a robust mechanism for handling end-of-file conditions.

- **Tests**
	- Added a suite of unit tests for the lexer to validate token recognition, whitespace handling, quote processing, operator recognition, and comment handling.

### CODE REVIEW

**Comment by alex216** [comment]
- [ ] `char *` -> `const char *` in `t_lexer`
- [ ] `get_next_token.c`: refactor
- [ ] `lexer.c`: use `TOKEN_EOF`
- [ ] add operator

**Review by rask24 — COMMENTED** [review]
plz check my comment @alex216

File: src/lexer/process_blank.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+	char	*token_value;
+
+	if (lexer->left == lexer->right)
+	{
+		discard_char(lexer);
+		return (NULL);
+	}
+	token_value = ft_substr(lexer->left, 0, lexer->right - lexer->left);
+	while (ft_isspace(*lexer->right))
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/process_blank.c]: done at 555303d Fix ft_isspace -> ft_isblank.

File: tests/unit/test_lexer.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
+#include "gtest/gtest.h"
+
+extern "C" {
+#include "lexer.h"
+}
+
+TEST(lexer, CheckTokenWord) {
+  const char *str = "ls";
```
[the remark this answers is not in the document]
Comment (alex216) [reply tests/unit/test_lexer.cpp:8]: Fixed at e3b1ab6 Rename variable: start -> left, end -> right, string -> input, starting -> left.

File: src/lexer/lexer_internal.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# define LEXER_INTERNAL_H
+
+# include "token.h"
+
+# define SINGLE_QUOTE '\''
+# define DOUBLE_QUOTE '\"'
+
+typedef struct s_lexer
+{
+	char		*left;
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/lexer_internal.h]: Fixed at 9573694 char * -> const char *.

File: src/lexer/lexer_internal.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	char		*right;
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/lexer_internal.h]: Fixed at
9573694 char * -> const char *

File: src/lexer/lexer.c [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include "token.h"
+
+t_token_list	*lexer(const char *input)
+{
+	t_token_list	*one_token;
+	t_token_list	*token_list;
+	t_lexer			lexer;
+
+	token_list = NULL;
+	construct_lexer((char *)input, &lexer);
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/lexer.c]: Fixed at
https://github.com/rask24/minishell/commit/9573694dd360eb83af0155e3c3e40f2a5c7c794d char * -> const char *

File: src/lexer/lexer_internal.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}				t_lexer;
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/lexer_internal.h]: decided to continue using `t_lexer` at cafe at 9/3.

File: src/lexer/get_next_token.c [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+t_token_list	*get_next_token(t_lexer *lexer)
+{
+	char	*token_value;
+
+	while (true)
+	{
+		if (*lexer->right == '\0')
+		{
+			token_value = ft_substr(lexer->left, 0, lexer->right - lexer->left);
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/get_next_token.c]: delete at refactoring process at
7668e28 Add process_eof.c file

File: .inputrc
[the remark this answers is not in the document]
Comment (alex216) [reply .inputrc:1]: Fixed at
2b69359 Fix delete .inputrc

File: src/lexer/constructor.c
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/constructor.c:1]: yes...?

File: tests/unit/test_lexer.cpp [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  t_token_type expected_token_array[] = {
+      TOKEN_WORD,          TOKEN_PIPE,          TOKEN_WORD,
+      TOKEN_DGREAT,        TOKEN_WORD,          TOKEN_DLESS,
+      TOKEN_LESS,          TOKEN_L_PARENTHESIS, TOKEN_WORD,
+      TOKEN_R_PARENTHESIS, TOKEN_R_PARENTHESIS, TOKEN_WORD,
+      TOKEN_OR_IF,         TOKEN_WORD,          TOKEN_EOF};
+
+  const char *expected_value_array[] = {"ls", "|",   "wc", ">>",  "whoami",
+                                        "<<", "<",   "(",  "ls",  ")",
+                                        ")",  "cat", "||", "echo"};
```
[the remark this answers is not in the document]
Comment (alex216) [reply tests/unit/test_lexer.cpp:204]: a2d6f52 Refactor test

File: tests/unit/test_lexer.cpp [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+  t_token_list *token = lexer(str);
+
+  int i = 0;
+  while (get_token_type(token) != TOKEN_EOF) {
+    EXPECT_EQ(get_token_type(token), expected_token_array[i]);
+    EXPECT_STREQ(get_token_value(token), expected_value_array[i]);
+    token = token->next;
+    i++;
+  }
```
[the remark this answers is not in the document]
Comment (alex216) [reply tests/unit/test_lexer.cpp:214]: a2d6f52 Refactor test

File: src/lexer/process_blank.c [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+/* ************************************************************************** */
+
+#include "lexer_internal.h"
+
+static void	_discard_blanks(t_lexer *lexer)
+{
+	while (ft_isblank(*lexer->right))
+	{
+		lexer->right++;
+	}
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/process_blank.c:20]: 60f8294 Refactor code by deleting {}

File: src/lexer/constructor.c [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+/*   Updated: 2024/08/31 14:46:12 by yliu             ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "lexer_internal.h"
+
+void	construct_lexer(const char *input, t_lexer *lexer)
+{
+	lexer->left = input;
+	lexer->right = input;
```
[the remark this answers is not in the document]
Comment (alex216) [reply src/lexer/constructor.c:18]: 91300d6 Initialize all field members

**Review by rask24 — COMMENTED** [review]
check the comment! @alex216

**Pull request merged**